src/error/Details.h

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (C) 1996-2023 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for detailss.
+ */
+
+#ifndef SQUID_SRC_ERROR_DETAILS_H
+#define SQUID_SRC_ERROR_DETAILS_H
+
+#include "error/Detail.h"
+#include "mem/PoolingAllocator.h"
+
+#include <vector>
+
+/// Multiple details of a single error, in "canonical" order without duplicates.
+/// Our "canonical" order is the approximate detail discovery order (e.g., the
+/// order of Error::update() calls). This class isolates multi-detail storage
+/// overheads from a common case of storing a single error detail.
+class ErrorDetails: public ErrorDetail
+{
+public:
+    /// Combines the already stored and the latest error details while
+    /// preserving uniqueness and canonical order. Each parameter may point to a
+    /// single detail or to an ErrorDetails object (with multiple details).
+    static void Merge(ErrorDetailPointer &storage, const ErrorDetailPointer &latest);
+
+    virtual ~ErrorDetails() = default;
+
+protected:
+    // use Merge() instead
+    ErrorDetails() = default;
+
+    /// adds the given detail unless we already have it
+    void mergeOne(const ErrorDetail &);
+
+    /// adds unique details (if any) from the given collection
+    void mergeMany(const ErrorDetails &);
+
+    /* ErrorDetail API */
+    virtual SBuf brief() const override;
+    virtual SBuf verbose(const HttpRequestPointer &) const override;
+
+private:
+    /// known unique details in canonical order
+    std::vector<ErrorDetailPointer, PoolingAllocator<ErrorDetailPointer> > details;
```

Comment:
TLDR: `std::vector` optimizes random access while `std::forward_list` makes such access very inefficient and awkward. This difference is real, but it does not automatically mean `std::forward_list` is a better choice for this private data member. We have to consider all factors. Availability of random access interfaces is a secondary concern here because the (private) code in question does not need to protect others from erroneous random access operations and erroneous random access inside this class is very unlikely. Performance concerns make `std::vector` a better choice overall.

> The definition you created for "canonical" forbids insertion or prepending entries,

As you know, both `vector` and `forward_list` support insertion of entries in arbitrary positions.  Canonical order makes such operations unnecessary in the current code. However, if you prefer "forbids", let's assume that interpretation for the sake of the argument.

> and reading in any order other than forward

That is not technically accurate: One could, in theory, meaningfully render events (stored in canonical order) in reverse, for example. However, again, if you prefer to think of this as a "read forward only" storage, let's assume that interpretation for the sake of the argument.

> `vector` violates all of those properties by default, `forward_list` only violates prepend by default. therefore it is clearly the better type.

Yes, `forward_list` is a slightly better type with respect to restricting access. However, it is worse type _overall_ in this context, and that is what matters. If we were talking about some publicly accessible storage where wrong access order would have been bad, the conclusion could have been different. Here, we are talking about a private data member in a small class that can and does exercise tight control over storage operations to prevent bad ones. In this case, performance matters more than (unwanted) availability of fast random access.

Reply:
Closing for now. We can open the topic of most optimal container type after the core meeting agreed redesign occurs.